app/components/primary_dashboard_community_activity_component/primary_dashboard_community_activity_component.html.erb

```diff
@@ -0,0 +1,69 @@
+<div
+  class="primary-dashboard-community-activity"
+  data-controller="primary-dashboard-community-activity-component"
+  data-primary-dashboard-community-activity-component-achievements-path-value="<%= achievements_path %>">
+
+  <% if @completed_activities.blank? && @incomplete_activities.blank? %>
+    <h2 class="govuk-heading-s">Complete at least one activity</h2>
+  <% end %>
+
+  <% if @completed_activities.present? %>
+    <div class="primary-dashboard-community-activity__activity" data-primary-dashboard-community-activity-component-target="activityDetails">
+      <h2 class="govuk-heading-s">Your completed <%= "activity".pluralize(@completed_activities.count) %></h2>
+      <% @completed_activities.each do |upa| %>
+        <div class="primary-dashboard-community-activity__activity-details">
+          <div class="primary-dashboard-community-activity__activity-details-header">
+            <div>
+              <%= image_pack_tag('media/images/icons/tick-green-dark.svg') %>
+            </div>
+            <div>
+              <p class="govuk-body-m govuk-!-margin-bottom-0"><%= upa.activity.title %></p>
+            </div>
+          </div>
+          <p class="govuk-body-s primary-dashboard-community-activity__activity-details-description"><%= upa.activity.public_copy_description.html_safe %></p>
+        </div>
+      <% end %>
+    </div>
+  <% end %>
+
+  <% if @incomplete_activities.present? %>
+    <div class="primary-dashboard-community-activity__activity" data-primary-dashboard-community-activity-component-target="activityDetails">
+      <h2 class="govuk-heading-s">Complete your chosen <%= "activity".pluralize(@incomplete_activities.count) %></h2>
+      <% @incomplete_activities.each do |upa| %>
+        <% achievement = (@current_user.achievements.find { |a| a.activity_id == upa.activity_id }) %>
```

Comment:
In here we pull out the achievement for the activity. But as the original query is from the user context, we could already have that in memory, saving the need for two iterations.

If you move the methods for complete and incomplete down to here, we can optimise the queries and reduce the need to look up the achievements.